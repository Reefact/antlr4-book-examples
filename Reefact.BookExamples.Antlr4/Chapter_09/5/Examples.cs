#region Usings declarations

using Antlr4.Runtime;
using Antlr4.Runtime.Misc;

using ApprovalTests;
using ApprovalTests.Reporters;

using NFluent;

using Xunit;

#endregion

namespace Reefact.BookExamples.Antlr4.Chapter_09._5 {

    [UseReporter(typeof(CustomReporter))]
    public class Examples {

        [Fact]
        public void wacky_sharp_lexical_error() {
            // Setup
            AntlrInputStream    inputStream   = AntlrInputStreamReader.Read("# class T { int i; }");
            IAntlrErrorStrategy errorStrategy = new BailErrorStrategy();
            // Exercise & verify
            Check.ThatCode(() => GRun.Read(inputStream, errorStrategy))
                 .Throws<GRunLexerException>()
                 .WhichMember(e => e.ErrorDetails)
                 .ContainsExactly("line 1:0 token recognition error at: '#'");
        }

        [Fact]
        public void missing_class_name_error() {
            // Setup
            AntlrInputStream    inputStream   = AntlrInputStreamReader.Read("class { int i; }");
            IAntlrErrorStrategy errorStrategy = new BailErrorStrategy();
            // Exercise & verify
            Check.ThatCode(() => GRun.Read(inputStream, errorStrategy))
                 .Throws<ParseCanceledException>();
        }

        [Fact]
        public void custom_error_message() {
            // Setup
            AntlrInputStream    inputStream   = AntlrInputStreamReader.Read("class T { int ; }");
            IAntlrErrorStrategy errorStrategy = new MyErrorStrategy();
            // Exercise
            GRun   grun   = GRun.Read(inputStream, errorStrategy);
            string output = grun.GetOutput();
            // Verify
            Approvals.Verify(output);
        }

    }

}