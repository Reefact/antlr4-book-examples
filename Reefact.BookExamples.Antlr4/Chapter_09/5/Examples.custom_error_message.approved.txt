line 1:14 Can't choose between alternatives
class T